lib/galaxy/objectstore/irods.py

```diff
@@ -731,8 +739,8 @@ def _get_filename(self, obj, **kwargs):
         #     if not os.path.exists(cache_path):
         #         os.makedirs(cache_path)
         #     return cache_path
-        # Check if the file exists in the cache first
-        if self._in_cache(rel_path):
+        # Check if the file exists in the cache first, always pull if file size in cache is zero
+        if self._in_cache(rel_path) and (dir_only or os.path.getsize(self._get_cache_path(rel_path)) > 0):
```

Comment:
@jdavcs - the reason the tests fail is in this line. 

I need to understand how it worked before. So, please correct me if I'm wrong. The way Galaxy works is that it creates a zero-size file in the cache for each output during job setup. I'm not really sure why it is needed, but it broke when I tried to skip that step. Then, if the job runs on the same machine or if Pulsar sends data directly to Galaxy, the cache file is replaced by real job output during set_metadata.py and everything is fine.

But, if the data from Pulsar job is sent to an object store, not to Galaxy, this file in cache on Galaxy side will still be 0 and never pulled from the object store. This combination is not tested by integration tests, I think. Has anybody ever tried that?

So, I had to add pulling to the cache if a file size if zero. But now irods gives an error since it tries to replace an existing file and irods options are not configured for that. In the previous version, get_filename skipped pulling from the cache if the file exists there, so it basically always skipped it.

Reply:
> Has anybody ever tried that?

it sure is a problem, I've tried to get rid of the empty file creation before the job starts but it wasn't straightforward (see https://github.com/galaxyproject/galaxy/pull/16494) so I put it on hold.

Reply:
@SergeyYakubov I was not aware of this problem, so thanks for the explanation! (I've learned something new).  Looking at the draft PR by Marius, I think your workaround is definitely good enough. The irods client not pulling from the cache is yet another issue.. 

Do you mind adding a one-liner comment to the irods client where you add the force flag option (line 448)? Just so it's clear to someone reading that code why that flag was necessary (you can also add a link to [your comment](https://github.com/galaxyproject/galaxy/pull/17025#discussion_r1394517033) for more context.